File: 1-2주차/11003.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const [n, l] = input[0].split(' ').map(Number);
const nums = input[1].split(' ').map(Number);

const deque = [];
const result = [];

for (let i = 0; i < n; i++) {
  const num = nums[i];

  while (deque.length > 0 && deque[deque.length - 1][0] > num) {
    deque.pop();
  }
  deque.push([num, i]);

  if (deque[0][1] <= i - l) {
    deque.shift();
  }
  result.push(deque[0][0]);
}

console.log(result.join(' '));

File: 1-2주차/11003.py
import sys
from collections import deque

input = sys.stdin.readline

n, l = map(int, input().split())
nums = list(map(int, input().split()))

dq = deque()
result = []

for i in range(n):
    num = nums[i]

    while dq and dq[-1][0] > num:
        dq.pop()
    dq.append((num, i))

    if dq[0][1] <= i - l:
        dq.popleft()

    result.append(dq[0][0])

print(' '.join(map(str, result)))
